### PR TITLE
Adds failed test tracking

### PR DESCRIPTION
**Pull request by rbuckton**
This PR adds the ability to track failing tests when building with the "runtests" or "runtests-parallel" build tasks. The names of failing tests are written to a _.failed-tests_ file in the workspace root. You can re-run failing tests using `jake runtests failed=true` (or `gulp runttests --failed`, once gulp is working again). 

Whenever a failing test passes, it is removed from the _.failed-tests_ file. If all tests pass, the _.failed-tests_ file is removed. This helps when running down a large number of failing tests, without having to continually rerun the entire test suite.

### CODE REVIEW

File: Gulpfile.js [hunk 1]
```diff
@@ -298,7 +298,8 @@ gulp.task(runJs, /*help*/ false, [typescriptServicesJs, tsserverlibraryDts], ()
 
 gulp.task(
     "tests",
-    "Builds the test infrastructure using the built compiler",
+        if (tests || runners || light || testTimeout || taskConfigsFolder) {
```
Comment (weswigham) [inline Gulpfile.js]: Is there a syntax error here?

File: scripts/run-failed-tests.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    if (waitForGrepValue) {
+        grep.push(arg.replace(/^"|"$/g, ""));
+        waitForGrepValue = false;
+        return;
+    }
+    args.push(arg);
+});
+
+let mocha = "./node_modules/mocha/bin/mocha";
```
Comment (weswigham) [inline scripts/run-failed-tests.js]: I think we should resolve this path using `__dirname` to handle if Jake or gulp is involved in a project subfolder.

Comment (rbuckton) [reply scripts/run-failed-tests.js]: Both jake and gulp reset cwd to the location of the jakefile/gulpfile.

Comment (weswigham) [reply scripts/run-failed-tests.js]: Alright, neat.

**Pull request merged**